app/controllers/concerns/checkout_steps.rb

```diff
@@ -39,9 +43,19 @@ def redirect_to_step
   def check_step
     case @order.state
     when "cart", "address", "delivery"
-      redirect_to checkout_step_path(:details) unless params[:step] == "details"
+      redirect_to checkout_step_path(:details) unless details_step?
     when "payment"
-      redirect_to checkout_step_path(:payment) if params[:step] == "summary"
+      redirect_to checkout_step_path(:payment) if summary_step?
     end
   end
+
+  def update_order_state
+    if @order.state == "confirmation" && payment_step?
+      @order.back_to_payment
+    end
+
+    return unless @order.state.in?(["payment", "confirmation"]) && details_step?
+
+    @order.back_to_address
+  end
```

Comment:
Looking good. A bit confusing though. I think that Rubocop's suggestion can make things worse.

The code path allows to first go back to payment and then also back to address. This path is impossible though because payment_step and details_step can't both be true. So we could return after back to payment which means that we don't have to think about the other conditions within the same request.

To make things worse, the edit action also redirects to the right step based on the order.

1. So first we check the order state and update the page (step).
2. And then depending on the page (step) we update the order state again. :confused:
3. And then we check the step again to redirect if we are not there. :upside_down_face: 

That looks like we could be doing at least one unnecessary redirect. So maybe instead of redirecting in `redirect_to_step_based_on_order` we should just find where we should go, update the order if necessary, and then redirect accordingly.

Actually, I now realise that I don't quite understand how it works. If the order state is confirmation and the step is payment then we trigger a redirect to summary before this method is called. There's no return though. So this code is then executed as well, setting the order back to payment. Now the order and the step are both in payment which means that the next check of the step doesn't detect any mismatch and doesn't redirect. So I guess that the redirect to summary is still executed. The summary page would then detect a mismatch and redirect back to payment. Okay, so this path works but we have a couple of useless redirects to where we were already. The controller spec doesn't assert on the response to the request. That would be better in a request spec.

Reply:
I cleaned up the code to make things clearer, see : https://github.com/openfoodfoundation/openfoodnetwork/pull/11677/commits/25af178011dbf084d0d264d583921d20e2a58655

> To make things worse, the edit action also redirects to the right step based on the order.
> 
>  1. So first we check the order state and update the page (step).
>  2. And then depending on the page (step) we update the order state again. 😕
>  3. And then we check the step again to redirect if we are not there. 🙃

Think you misread the code here, 1. and 3. are mutually exclusive. 1. happens if `params[:step]` isn't set and 3. happens if `params[:step]` is set.